code/datums/state_machine/state.dm

```diff
@@ -0,0 +1,36 @@
+// An individual state, defined as a `/decl` to save memory.
+// On a directed graph, these would be the nodes themselves, connected to each other by unidirectional arrows.
+/decl/state
+	// Transition decl types, which get turned into refs to those types.
+	// Note that the order DOES matter, as decls earlier in the list have higher priority
+	// if more than one becomes 'open'.
+	var/list/transitions = null
+
+// Called by `SSstate_machines` to have the state decls build the directed graph themselves.
+/decl/state/proc/set_up()
+	if(!LAZYLEN(transitions))
+		return
+	for(var/i in 1 to transitions.len)
+		var/decl/state_transition/T = GET_DECL(transitions[i])
+		if(!istype(T))
+			CRASH("[type] expected a state transition object but was not given one.")
+		T.set_up(src)
+		transitions[i] = T // Replace the path with a ref to the instance.
+
+// Returns a list of transitions that a FSM could switch to.
+// Note that `holder` is NOT the FSM, but instead the thing the FSM is attached to.
+/decl/state/proc/get_open_transitions(datum/holder)
+	. = list()
+	for(var/thing in transitions)
```

Comment:
You can do `for(var/decl/state_transition/T as anything in transitions)` for the performance increase of an untyped loop without having to cast the value afterwards.